### PR TITLE
Generic LoadTime Optimizations

### PR DESCRIPTION
**Pull request by GlebKazantaev**
**Description:**
This PR includes findings from https://github.com/openvinotoolkit/openvino/pull/3202 to optimize transformations pipeline execution.

**Details:**
* Updated AlgebraicSimplification and NopElimination passes to use GraphRewrite/MatcherPass base classes.
* Added ability to register existing class based on GraphRewrite inside GraphRewrite anchor.
* Refactored CommonOptimizations pipeline to avoid excess graph traversals and Validation passes calls.
* Updated return value to false for container like transformations to avoid excess Validation passes.
* Disabled nodes validation in ConstantFolding pass for cases when graph is not modified.

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp [hunk 1]
```diff
@@ -234,31 +235,32 @@ static bool replace_transpose_with_reshape(shared_ptr<Node> transpose) {
...
+NAME() { \
+    MATCHER_SCOPE(NAME);    \
+    auto match_node = ngraph::pattern::wrap_type<OP>(); \
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) { \
+        return FUNC(m.get_match_root()); \
+    }; \
+    auto m = std::make_shared<ngraph::pattern::Matcher>(match_node, matcher_name); \
+    register_matcher(m, callback); \
+}  \
+}; \
```
Comment (ilya-lavrenov) [inline inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp:253]: maybe we can use templates? E.g. create template class and instantiate with different ops and functions:
```c++
template <typename Op, typename Func>
class SimpleMatcher { .. };

ngraph::pass::AlgebraicSimplification::AlgebraicSimplification() {
    add_matcher<SimpleMatcher<opset3::Gather, simplify_gather>>();
    add_matcher<SimpleMatcher<opset2::ShapeOf, simplify_gather_shapeof>>();
    add_matcher<SimpleMatcher<opset3::ShapeOf, simplify_gather_shapeof>>();
    add_matcher<SimpleMatcher<opset3::Transpose, replace_transpose_with_reshape>>();
}
```

BTW, `Func` can be passed as a constructor argument.

File: inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 2]
```diff
@@ -327,33 +330,38 @@ static bool eliminate_squeeze(const std::shared_ptr<Node>& node) {
...
-            clobbered |= handler->second(node);
-        }
-    }
+SIMPLE_MATCHER_PASS_DEFINITION(EliminatePad, opset3::Pad, eliminate_nop);
+SIMPLE_MATCHER_PASS_DEFINITION(EliminateConvert, opset3::Convert, eliminate_convert);
+SIMPLE_MATCHER_PASS_DEFINITION(EliminateReshape, opset3::Reshape, eliminate_reshape_v1);
+SIMPLE_MATCHER_PASS_DEFINITION(EliminateConcat, opset3::Concat, eliminate_concat);
+SIMPLE_MATCHER_PASS_DEFINITION(EliminateSqueeze, opset3::Squeeze, eliminate_squeeze);
+SIMPLE_MATCHER_PASS_DEFINITION(EliminateUnsqueeze, opset3::Unsqueeze, eliminate_unsqueeze);
+SIMPLE_MATCHER_PASS_DEFINITION(EliminateBroadcast, op::v1::Broadcast, eliminate_nop);
```
Comment (ilya-lavrenov) [inline inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp:357]: Looks looks `SimpleMatcher` can be put into some `src/simple_matcher.hpp`

File: inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp (hunk 1 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp:253]: Anyway we will have to define NGRAPH_RTTI_DEFINITION for each specialization and also this passes (AlgebraicSimplification and NopElimination) will be refactored (Ticket: 37380) bun in a separate PR.

 The changes I've made are temporary but they are necessary to have this passes be united into single graph traversal.

File: inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp (hunk 2 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp:357]: I believe that after refactoring that I mentioned above this SimpleMatcher won't be needed. Anyway we will see it in another PR with AlgebraicSimplification and NopElimination refactoring.

Comment (ilya-lavrenov) [reply inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp:357]: Ok

**Comment by GlebKazantaev** [comment]
Waiting for internal performance and accuracy validation.

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by GlebKazantaev** [comment]
Internal validation successfully passed.

**Pull request merged**